File: Data Structures & Algorithms/singlyLinkedList/submission-6.js
/**
 * Singly Linked List Node
 */
class ListNode {
    /**
     * @param {number} val - Value of the node
     * @param {ListNode} [nextNode=null] - Reference to the next node
     */
    constructor(val = -1, nextNode = null) {
        this.val = val;
        this.next = nextNode;
    }
}

class LinkedList {
    constructor(val = 0, next = null) {
        this.head = new ListNode();
        this.tail = this.head;
    }

    /**
     * @param {number} index
     * @return {number}
     */
    get(index) {
        let curr = this.head.next;
        let i = 0;
        while (curr) {
            if (i === index) {
                return curr.val
            }
            curr = curr.next;
            i++;
        }

        return -1
    }

    /**
     * @param {number} val
     * @return {void}
     */
    insertHead(val) {
        const node = new ListNode(val)   
        node.next = this.head.next;
        this.head.next = node;
        if(!node.next) {
            this.tail = node;
        }

    }

    /**
     * @param {number} val
     * @return {void}
     */
    insertTail(val) {
        const node = new ListNode(val)
        this.tail.next = node;
        this.tail = this.tail.next;
    }

    /**
     * @param {number} index
     * @return {boolean}
     */
    remove(index) {
        let i = 0;
        let curr = this.head;
        while (i < index && curr) {
            i++;
            curr = curr.next;
        }

        if (curr && curr.next) {
            if (curr.next === this.tail) {
                this.tail = curr;
            }
            curr.next = curr.next.next;
            return true;
        }
        return false;
    }

    /**
     * @return {number[]}
     */
    getValues() {
        let curr = this.head.next;
        const values = [];
        while (curr) {
            values.push(curr.val)
            curr = curr.next;
        }

        return values;
    }
}

File: Data Structures & Algorithms/lru-cache/submission-6.py
class Node:
    def __init__(self, key=-1, val=-1):
        self.key = key
        self.val = val
        self.prev = None
        self.next = None


class LRUCache:

    def __init__(self, capacity: int):
        self.capacity = capacity
        self.size = 0
        self.nodes = {}

        self.head = Node()
        self.tail = Node()
        self.head.next = self.tail
        self.tail.prev = self.head


    def get(self, key: int) -> int:
        if key not in self.nodes:
            return -1

        node = self.nodes[key]
        value = node.val
        self._remove(key)
        self._insert(key, value)

        return value


    def put(self, key: int, value: int) -> None:
        if key in self.nodes:
            self._remove(key)

        if self.size == self.capacity:
            self._remove(self.head.next.key)

        self._insert(key, value)
    

    def _insert(self, key, value) -> None:
        node = Node(key,value)
        node.prev = self.tail.prev
        node.next = self.tail
        self.tail.prev.next = node
        self.tail.prev = node

        if key not in self.nodes:
            self.size += 1

        self.nodes[key] = node


    def _remove(self, key: int) -> bool:
        if key not in self.nodes:
            return False

        node = self.nodes[key]
        node.prev.next = node.next
        node.next.prev = node.prev
        node.prev = None
        node.next = None

        self.size -= 1
        del self.nodes[key]
        
        return True
